src/main/java/com/minecolonies/entity/ai/citizen/guard/AbstractEntityAIGuard.java

```diff
@@ -0,0 +1,325 @@
+     */
+    private static final int START_SEARCH_DISTANCE = 5;
+
+    /**
+     * Basic delay after operations.
+     */
+    private static final int BASE_DELAY = 1;
+
+    /**
+     * Max amount the guard can shoot arrows before restocking.
+     */
+    private static final int MAX_ATTACKS = 50;
+
+    /**
+     * Y range in which the guard detects other entities.
+     */
+    private static final double HEIGHT_DETECTION_RANGE = 10D;
+
+    /**
+     * Path that close to the patrol target.
+     */
+    private static final int PATH_CLOSE = 3;
+    /**
+     * The current target.
+     */
+    protected EntityLivingBase targetEntity;
+    /**
+     * Amount of arrows already shot or sword hits dealt.
+     */
+    protected int attacksExecuted = 0;
+    /**
+     * The distance the guard is searching entities in currently.
+     */
+    private int currentSearchDistance = START_SEARCH_DISTANCE;
+    /**
+     * Current goTo task.
+     */
+    private BlockPos currentPatrolTarget;
+    /**
+     * Containing all close entities.
+     */
+    private List<Entity> entityList;
+
+    /**
+     * Sets up some important skeleton stuff for every ai.
+     *
+     * @param job the job class
+     */
+    protected AbstractEntityAIGuard(@NotNull final JobGuard job)
+    {
+        super(job);
+        super.registerTargets(
+          new AITarget(IDLE, () -> START_WORKING),
+          new AITarget(START_WORKING, () -> GUARD_RESTOCK)
+        );
+    }
+
+    /**
+     * Can be overridden in implementations.
+     * <p>
+     * Here the AI can check if the armour have to be re rendered and do it.
+     */
+    @Override
+    protected void updateRenderMetaData()
+    {
+        updateArmor();
+    }
+
+    /**
+     * Updates the equipment. Always take the first item of each type and set it.
+     */
+    protected void updateArmor()
+    {
+        worker.setItemStackToSlot(EntityEquipmentSlot.CHEST, null);
+        worker.setItemStackToSlot(EntityEquipmentSlot.FEET, null);
+        worker.setItemStackToSlot(EntityEquipmentSlot.HEAD, null);
+        worker.setItemStackToSlot(EntityEquipmentSlot.LEGS, null);
+
+        for (int i = 0; i < worker.getInventoryCitizen().getSizeInventory(); i++)
+        {
+            ItemStack stack = worker.getInventoryCitizen().getStackInSlot(i);
+
+            if (stack == null || stack.stackSize == 0)
+            {
+                worker.getInventoryCitizen().setInventorySlotContents(i, null);
+                continue;
+            }
+
+            if (stack.getItem() instanceof ItemArmor && worker.getItemStackFromSlot(((ItemArmor) stack.getItem()).armorType) == null)
+            {
+                worker.setItemStackToSlot(((ItemArmor) stack.getItem()).armorType, stack);
+            }
+        }
+    }
+
+    /**
+     * Goes back to the building and tries to take armour from it when he hasn't in his inventory.
+     *
+     * @return the next state to go to.
+     */
+    protected AIState goToBuilding()
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "goToBuilding" is 11 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)